In [1]:
##import libraries
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
##Set Path
path = r'/Users/heatherstewart/Dropbox/Mac/Desktop/Instacart Basket analysis 3'

In [3]:
path

'/Users/heatherstewart/Dropbox/Mac/Desktop/Instacart Basket analysis 3'

In [4]:
##import product csv file
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'),index_col = False)

In [5]:
##import orders csv file
df_ords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data','orders_wrangled.csv'),index_col = False)

In [6]:
df_ords.head()

Unnamed: 0  user_id  order_category_number  order_day_of_week  \
0           0        1                      1                  2   
1           1        1                      2                  3   
2           2        1                      3                  3   
3           3        1                      4                  4   
4           4        1                      5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [7]:
df_ords.shape

(3421083, 6)

In [8]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [9]:
df_ords = df_ords.drop('Unnamed: 0', axis = 1)

In [10]:
df_ords.head()

user_id  order_category_number  order_day_of_week  order_hour_of_day  \
0        1                      1                  2                  8   
1        1                      2                  3                  7   
2        1                      3                  3                 12   
3        1                      4                  4                  7   
4        1                      5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [11]:
# Create a dataframe
df_test = pd.DataFrame()

In [12]:
# Create a mixed typed column
df_test['mix'] = ['a','b',1, True]

In [13]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [15]:
# Check for weird mix types using "for loop"

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [16]:
# Locating missing observations in df_prods dataframe using 'isnull'
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
# To view the missing observation create a subset(new dataframe) of the dataframe containing only the value in question.

df_nan = df_prods [df_prods['product_name'].isnull()==True]

In [18]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [19]:
# Viewing the df_prods dataframe to compare number of rows

df_prods.shape

(49693, 5)

In [20]:
##Creating a new dataframe to drop the missing (NaN) values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [22]:
df_prods_clean.shape

(49677, 5)

In [23]:
# Looking for duplicates in the df_prods_clean dataframe

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [24]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [25]:
# Checking rows in the df_prods_clean dataframe

df_prods_clean.shape

(49677, 5)

In [26]:
#Creating a new dataframe that doesn't include the duplicates identified using the 'drop_duplicates()'function.

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [27]:
df_prods_clean_no_dups.shape

(49672, 5)

In [30]:
#Exporting dataframe into "Prepared Data" folder

df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data','product_checked.csv'))

## PERFORMING DATA CONSISTENCY 4.5 EXERCISE

In [31]:
#STEP 2:Run "describe" function on df_ords dataframe and share any inconsistency in a markdown cell.

df_ords.describe()

user_id  order_category_number  order_day_of_week  \
count  3.421083e+06           3.421083e+06       3.421083e+06   
mean   1.029782e+05           1.715486e+01       2.776219e+00   
std    5.953372e+04           1.773316e+01       2.046829e+00   
min    1.000000e+00           1.000000e+00       0.000000e+00   
25%    5.139400e+04           5.000000e+00       1.000000e+00   
50%    1.026890e+05           1.100000e+01       3.000000e+00   
75%    1.543850e+05           2.300000e+01       5.000000e+00   
max    2.062090e+05           1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

##The column "order_hour_of_day" max is showing 2.3 and need to be investigated as there is 24 hours max in a day. 

In [34]:
## Step 3: Check for mix-type data in df_ords dataframe. 

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

##Step 4: There were no mix-type data found in the df_ords dataframe.

In [35]:
## Step 5: Checking for missing values in df_ords dataframe. 

df_ords.isnull().sum()

user_id                        0
order_category_number          0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

##The column "days_since_prior_order has 206209 missing values. The missing values can be missing due non inputation at the start of data collection. 

In [36]:
## Step 6: Addressing the missing values for df_ords dataframe by using the mean as a replacement value. 

df_ords['days_since_prior_order'].fillna(1.11148, inplace = True)

In [38]:
df_ords.isnull().sum()

user_id                   0
order_category_number     0
order_day_of_week         0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

 ##I choose to use the mean as a replacement value for imputation as there is signigicant amount of value missing and dropping them could affect my analysis. The missing values have a higher chance of being closer in value to mean based on the principle of missing values being closer to the mean if they were not missing. 

In [40]:
##Step 7: Checking for duplicates in the df_ords dataframe

df_dups = df_ords[df_ords.duplicated()]

In [41]:
df_dups

Empty DataFrame
Columns: [user_id, order_category_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## df_dataframe returned no duplicated values. 

In [42]:
##Exporting df_ords dataframe 

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data','orders_checked.csv'))